src/vm/dllimport.cpp

```diff
@@ -6342,11 +6343,81 @@ NATIVE_LIBRARY_HANDLE NDirect::LoadLibraryModuleViaHost(NDirectMethodDesc * pMD,
     args[ARGNUM_1]  = PTR_TO_ARGHOLDER(ptrManagedAssemblyLoadContext);
 
     // Make the call
-    CALL_MANAGED_METHOD(hmod,LPVOID,args);
+    CALL_MANAGED_METHOD(hmod, NATIVE_LIBRARY_HANDLE, args);
 
     GCPROTECT_END();
 
-    return (NATIVE_LIBRARY_HANDLE)hmod;
+    return hmod;
+}
+
+// Return the AssemblyLoadContext for an assembly
+INT_PTR GetManagedAssemblyLoadContext(Assembly* pAssembly)
+{
+    STANDARD_VM_CONTRACT;
+
+    PTR_ICLRPrivBinder pBindingContext = pAssembly->GetManifestFile()->GetBindingContext();
+    if (pBindingContext == NULL)
+    {
+        // GetBindingContext() returns NULL for System.Private.CoreLib
+        return NULL;
+    }
+
+    UINT_PTR assemblyBinderID = 0;
+    IfFailThrow(pBindingContext->GetBinderID(&assemblyBinderID));
+
+    ICLRPrivBinder *pCurrentBinder = reinterpret_cast<ICLRPrivBinder *>(assemblyBinderID);
+    ICLRPrivBinder *pTPABinder = GetAppDomain()->GetTPABinderContext();
+    BOOL isDefaultContext = AreSameBinderInstance(pCurrentBinder, pTPABinder);
+
+    INT_PTR ptrManagedAssemblyLoadContext = (isDefaultContext) ?
```

Comment:
Wow. This is really confusing and looks like a bug. Basically we get two seemingly different interfaces and see if they are the same binder instance. Based on that information we then cast only the `pCurrentBinder` to one of two different types. I have no idea if this is a bug or not, but it is non-obvious and very dubious as an intuitive architecture. Clearly it may just be how it is, but can we have a comment describing the logic here?

Reply:
The code here deals with two implementations of the same `ICLRPrivBinder` interface: `CLRPrivBinderCoreCLR` for the TPA binder in the `default ALC`, and `CLRPrivBinderAssemblyLoadContext` for all other `ALC`s.  So, we check if the current-binder is the TPA binder, and treat the two implementations `ICLRPrivBinder`  differently to obtain the associated `ALC` handle. 

I can add the above explanation as comments in this case, but this logic exists in several places. For example: [1](https://github.com/dotnet/coreclr/blob/ee8bc3272391e46862dd8d27d3e086dc3bf953cf/src/vm/assemblynative.cpp#L203) [2](https://github.com/dotnet/coreclr/blob/eff427c02a89e135c38a96032feb3c9a6a13cf5b/src/vm/dllimport.cpp#L6307)

I'll also add a check to exclude `CLRPrivBinderWinRT` contexts similar to [this](https://github.com/dotnet/coreclr/blob/eff427c02a89e135c38a96032feb3c9a6a13cf5b/src/vm/dllimport.cpp#L6314) code.